### PR TITLE
Add Patch as an input/output option with UI by zmarotrix. 

### PR DESCRIPTION
**Pull request by ZoeyZolotova**
Add Patch as an out option.
Add button to apply a patch.

### CODE REVIEW

File: Forms/MainForm.cs [hunk 1]
```diff
@@ -129,6 +131,15 @@ private void bRandomise_Click(object sender, EventArgs e)
             bgWorker.RunWorkerAsync();
         }
 
+        private void bApplyPatch_Click(object sender, EventArgs e)
+        {
+            if (openPatch.ShowDialog() == DialogResult.OK)
+            {
+                _settings.InputPatchFilename = openPatch.FileName;
+                bRandomise_Click(sender, e);
```
Comment (Kjelli) [inline Forms/MainForm.cs]: Minor; I'd rather not call the click function here, but instead have both the `bRandomise_Click()` and `bApplyPatch_Click()` call a common function.

File: Utils/ItemSwapUtils.cs [hunk 2]
```diff
@@ -137,9 +137,12 @@ public static void WriteNewBottle(int ItemSlot, int NewItem)
             for (int i = 0; i < RomData.BottleIndices[ItemSlot].Length; i++)
             {
                 int offset = RomData.BottleIndices[ItemSlot][i] * 6 + baseaddr;
-                fileData[offset + 3] = RomData.BottleList[NewItem][0].ItemGained;
-                fileData[offset + 4] = RomData.BottleList[NewItem][0].Index;
-                fileData[offset + 5] = RomData.BottleList[NewItem][0].Message;
+                fileData.Write(offset + 3, new byte[]
+                {
+                    RomData.BottleList[NewItem][0].ItemGained,
```
Comment (Kjelli) [inline Utils/ItemSwapUtils.cs]:
```csharp
var newBottle = RomData.BottleList[NewItem][0];
fileData.Write(offset + 3, new byte[]
                {
                    newBottle.ItemGained,
                    ....
```

File: Utils/ObjUtils.cs [hunk 3]
```diff
@@ -11,15 +12,15 @@ public static int GetObjSize(int obj)
         {
             int f = RomUtils.GetFileIndexForWriting(Addresses.ObjTable);
             int basea = Addresses.ObjTable - RomData.MMFileList[f].Addr;
-            return (int)(ReadWriteUtils.Arr_ReadU32(RomData.MMFileList[f].Data, basea + (obj * 8) + 4)
-                - ReadWriteUtils.Arr_ReadU32(RomData.MMFileList[f].Data, basea + (obj * 8)));
+            return (int)(ReadWriteUtils.Arr_ReadU32(RomData.MMFileList[f].Data.ReadonlyData, basea + (obj * 8) + 4)
```
Comment (Kjelli) [inline Utils/ObjUtils.cs]:
```csharp
var fileData = RomData.MMFileList[f].Data.ReadonlyData;
return (int)(ReadWriteUtils.Arr_ReadU32(fileData, basea + (obj * 8) + 4)
                - ReadWriteUtils.Arr_ReadU32(fileData, basea + (obj * 8)))
```

**Comment by mzxrules** [comment]
I don't like the way that this is implemented. The biggest reason is that we don't need to track the rom modifications as we are building it, we can just focus on building the rom first, and then compare the file tables of the base and final rom to compute the diffs. The secondary issue is that this further solidifies the bad design in the way MMR handles files by adding another complication

**Pull request merged**